unit/solvers/refinement/bv_refinement/add_lemma.cpp

```diff
@@ -0,0 +1,136 @@
+/*******************************************************************\
+
+ Module: Unit tests for bv refinement
+
+ Author: DiffBlue Limited. All rights reserved.
+
+\*******************************************************************/
+
+#include <testing-utils/catch.hpp>
+
+#include <util/arith_tools.h>
+#include <util/std_types.h>
+#include <util/std_expr.h>
+#include <solvers/refinement/bv_refinement.h>
+#include <solvers/sat/satcheck.h>
+#include <util/symbol_table.h>
+#include <java_bytecode/java_types.h>
+#include <util/config.h>
+#include <iostream>
+
+SCENARIO("add lemma", "[core][solvers][refinement][bv_refinement]")
+{
+  // Without this line, the program aborts with:
+  // `Reason: pointer must have non-zero width`
+  config.set_arch("none");
+
+  GIVEN("A solver")
+  {
+    satcheck_no_simplifiert sat_check;
+    symbol_tablet symbol_table;
+    namespacet ns(symbol_table);
+    bv_refinementt::infot info;
+    info.ns = &ns;
+    info.prop = &sat_check;
+    info.refine_arithmetic = true;
+    info.refine_arrays = true;
+    info.max_node_refinement = 5;
+    info.ui = ui_message_handlert::uit::PLAIN;
+    bv_refinementt solver(info);
+
+    WHEN("We give two contradictory inputs")
+    {
+      const symbol_exprt s1("symbol_var1", java_int_type());
+      const exprt constant_int = from_integer(99, java_int_type());
+      const binary_relation_exprt lemma1(s1, ID_ge, constant_int);
+      const not_exprt lemma2(lemma1);
+
+      solver << lemma1;
+      solver << lemma2;
+
+      THEN("it should be unsatisfiable")
+      {
+        const auto result = solver();
+        REQUIRE(result == decision_proceduret::resultt::D_UNSATISFIABLE);
+      }
```

Comment:
Yes but this doesn't appear to test bv_refinement...

Reply:
Oh yes right, `boolbvt` is enough to solve this. I added tests for different solvers. It seems the example works with `boolbvt` and `bv_pointerst` but crashes with `bv_refinementt`

Reply:
The crash should be fixed; especially if it works without refinement.